### PR TITLE
log/Log.cc: speed optimizations. Merged 3 writes into 1.

### PR DESCRIPTION
**Pull request by aclamk**
Got rid of std::string construction.
More unification on syslog,stderr,fd.

Signed-off-by: Adam Kupczyk <akupczyk@mirantis.com>

### CODE REVIEW

File: src/log/Log.cc [hunk 1]
```diff
@@ -227,25 +227,23 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
...
-      if (do_fd) {
-	int r = safe_write(m_fd, buf, buflen);
-	if (r >= 0)
-	  r = safe_write(m_fd, s.data(), s.size());
-	if (r >= 0)
-	  r = write(m_fd, "\n", 1);
-	if (r < 0)
-	  cerr << "problem writing to " << m_log_file << ": " << cpp_strerror(r) << std::endl;
-      }
+      buflen += e->snprintf(buf + buflen, sizeof(buf) - buflen - 1 );
```
Comment (liewegas) [inline src/log/Log.cc]: Good catch!  I bet we can do even better with something that looks like pwritev(2)... then we avoid doing a memcpy into buf entirely, and instead construct an iovec array (there should be a max of 3 or 4 distinct inputs).  The write or pwritev syscall is doing another copy anyway, so the one into buf shouldn't be necessary.

Comment (aclamk) [inline src/log/Log.cc]: I have tried using writev but it is actually slower.
Below using writev vs using memcpy to make 1 buffer.
size=  6 IOVS=3 time=107768usec
size=  6 WRITE  time=84712usec
size= 12 IOVS=3 time=131376usec
size= 12 WRITE  time=120907usec
size= 24 IOVS=3 time=183488usec
size= 24 WRITE  time=172170usec
size= 48 IOVS=3 time=271790usec
size= 48 WRITE  time=271460usec
size= 96 IOVS=3 time=674332usec
size= 96 WRITE  time=424097usec
size=192 IOVS=3 time=769107usec
size=192 WRITE  time=784972usec
size=384 IOVS=3 time=2253296usec
size=384 WRITE  time=1509636usec
size=768 IOVS=3 time=2977487usec
size=768 WRITE  time=2974786usec

Comment (yuyuyu101) [inline src/log/Log.cc]: It make sense to me that very small vector will result in bad performance. I guess most of logs should be small enough

**Comment by liewegas** [comment]
Oops.. this passed testing but I merged too soon (and reverted).  Two issues:

1- This is putting lot sof \0 in the log output.  View a resulting log with 'less' to see.
2- Misc style.. mostly missing whitespace between if and (, on either side of =.  Please see CodingStyle.

Thanks!

**Pull request merged**